src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -301,25 +356,50 @@ public void onMessage(ConsumerRecord<K, V> record, Acknowledgment acknowledgment
 			Message<?> message = null;
 			try {
 				message = toMessagingMessage(record, acknowledgment);
+				setAttributesIfNecessary(record, message);
 			}
 			catch (RuntimeException e) {
-				Exception exception = new ConversionException("Failed to convert to message for: " + record, e);
-				if (getErrorChannel() != null) {
-					getMessagingTemplate().send(getErrorChannel(), new ErrorMessage(exception));
-				}
+				RuntimeException exception = new ConversionException("Failed to convert to message for: " + record, e);
+				sendErrorMessageIfNecessary(null, exception);
 			}
 			if (message != null) {
-				sendMessage(message);
+				try {
+					sendMessage(message);
+				}
+				finally {
+					attributesHolder.remove();
```

Comment:
Any comments here?
And I think we still need something for the `EnchancedErrorMessage` in the SI-4.3.10 to support this PR.
Thanks